cmd/gen-go-sample/main.go

```diff
@@ -0,0 +1,227 @@
+			log.Fatal(errors.E(err, "error reading GAPIC config file"))
+		}
+		donec <- struct{}{}
+	}()
+	go func() {
+		descBytes, err := ioutil.ReadFile(*descFname)
+		if err != nil {
+			log.Fatal(errors.E(err, "cannot read proto descriptor file"))
+		}
+
+		if err := proto.Unmarshal(descBytes, &gen.desc); err != nil {
+			log.Fatal(errors.E(err, "error reading proto descriptor file"))
+		}
+
+		gen.dinfo = pbinfo.Of(gen.desc.GetFile())
+		donec <- struct{}{}
+	}()
+
+	<-donec
+	<-donec
+
+	for _, iface := range gen.gapic.Interfaces {
+		for _, meth := range iface.Methods {
+			valSets := map[string]GapicValueSet{}
+			for _, vs := range meth.SampleValueSets {
+				valSets[vs.ID] = vs
+			}
+
+			for _, sam := range meth.Samples.Standalone {
+				for _, vsID := range sam.ValueSets {
+					vs, ok := valSets[vsID]
+					if !ok {
+						log.Fatal(errors.E(nil, "value set not found: %q", vsID))
+					}
+
+					gen.reset()
+					if err := gen.genSample(iface.Name, meth.Name, vs); err != nil {
+						log.Fatal(errors.E(err, "value set: %q", vs))
+					}
+					if err := gen.commit(!*nofmt); err != nil {
+						log.Fatal(errors.E(err, "value set: %q", vs))
+					}
+				}
+			}
+		}
+	}
+}
+
+type generator struct {
+	desc  descriptor.FileDescriptorSet
+	dinfo pbinfo.Info
+	gapic GapicConfig
+
+	pt   printer.P
+	imps map[pbinfo.ImportSpec]bool
+}
+
+func (g *generator) reset() {
+	g.pt.Reset()
+	for imp := range g.imps {
+		delete(g.imps, imp)
+	}
+}
+
+func (g *generator) commit(gofmt bool) error {
+	// We'll gofmt unless user asks us to not, so no need to think too hard about sorting
+	// "correctly". We just want a deterministic output here.
+	var imps []pbinfo.ImportSpec
+	for imp := range g.imps {
+		imps = append(imps, imp)
+	}
+	sort.Slice(imps, func(i, j int) bool {
+		if imps[i].Path != imps[j].Path {
+			return imps[i].Path < imps[j].Path
+		}
+		return imps[i].Name < imps[j].Name
+	})
+
+	var file bytes.Buffer
+	file.WriteString("package main\n")
+	file.WriteString("import(\n")
+	for _, imp := range imps {
+		fmt.Fprintf(&file, "%s %q", imp.Name, imp.Path)
+	}
+	file.WriteString(")\n")
+	file.Write(g.pt.Bytes())
+
+	b := file.Bytes()
+	if gofmt {
+		b2, err := format.Source(b)
+		if err != nil {
+			return err
+		}
+		b = b2
+	}
+	_, err := os.Stdout.Write(b)
+	return err
+}
+
+func (g *generator) genSample(ifaceName, methName string, valSet GapicValueSet) error {
+	serv := g.dinfo.Serv["."+ifaceName]
+	if serv == nil {
+		return errors.E(nil, "can't find service: %q", ifaceName)
+	}
+
+	servSpec, err := g.dinfo.ImportSpec(serv)
+	if err != nil {
+		return errors.E(err, "can't import service: %q", ifaceName)
+	}
+
+	var meth *descriptor.MethodDescriptorProto
+	for _, m := range serv.GetMethod() {
+		if m.GetName() == methName {
+			meth = m
+			break
+		}
+	}
+	if meth == nil {
+		return errors.E(nil, "service %q doesn't have method %q", serv.GetName(), methName)
+	}
+
+	p := g.pt.Printf
+
+	p("// [START %s]", valSet.ID)
```

Comment:
The region tag comes from the sample spec, not from the value set ID. If you're deferring this, add a TODO.

Reply:
Done